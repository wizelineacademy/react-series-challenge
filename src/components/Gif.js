import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { toggleFavorite } from '../actions/favorites';
import { isFavorited } from '../selectors/favorites';
import { selectedGifChange } from '../actions/selected';

const FAVORITE_ICON = '⭐';

export const StyledGif = styled.img`
  width: 100%;
  height: auto;
  display: block;
`;

export const StyledFavoritedIcon = styled.a`
  cursor: pointer;
  transition: filter .15s ease;
  content: '${FAVORITE_ICON}';
  position: absolute;
  bottom: 10px;
  right: 10px;
  ${({ active }) => active ? '' : `filter: grayscale(100%);`}

  &:hover: {
    text-shadow: -1px 0px 4px rgba(20, 20, 20, 1);
  }
`;

export const Gif = ({ big, gif, favorited, selectedGifChange, toggleFavorite }) => {
  const src = !big
    ? gif.images.fixed_width_downsampled.url
    : gif.images.original.url;
  return (
    <React.Fragment>
      <StyledFavoritedIcon onClick={() => toggleFavorite(gif)} active={favorited}>
        {FAVORITE_ICON}
      </StyledFavoritedIcon>
      <Link to={`?gifId=${gif.id}`}>
        <StyledGif src={src} onClick={() => selectedGifChange(gif)} />
      </Link>
    </React.Fragment>
  );
};

const mapStateToProps = (state, { gif }) => ({
  favorited: isFavorited(gif.id, state)
});

const mapDispatchToProps = {
  toggleFavorite,
  selectedGifChange
};

export default connect(mapStateToProps, mapDispatchToProps)(Gif);
