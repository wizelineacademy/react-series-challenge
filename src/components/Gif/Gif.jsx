import React from 'react';
import PropTypes from 'prop-types';

import {
  GifContainer,
  GifTitle,
  GifImage,
  GifLikeContainer,
} from './Gif.style';

import likeContainer from './favorite-heart-like-container.png';
import likeSelected from './favorite-heart-like-selected.png';

const Gif = ({ object, addFavorite, removeFavorite }) => (
  <GifContainer>
    <GifTitle>{object.title}</GifTitle>
    <GifImage src={object.embed_url} title={object.title} />
    {
      object.isFavorite ?
        (
          <GifLikeContainer
            src={likeSelected}
            alt="Remove from favorite"
            onClick={() => removeFavorite({ payload: { gif: object } })}
          />
        ) :
        (
          <GifLikeContainer
            src={likeContainer}
            alt="Add to favorite"
            onClick={() => addFavorite({ payload: { gif: object } })}
          />
        )
    }
  </GifContainer>
);

Gif.propTypes = {
  object: PropTypes.shape({
    embed_url: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    isFavorite: PropTypes.bool.isRequired,
  }).isRequired,
  addFavorite: PropTypes.func.isRequired,
  removeFavorite: PropTypes.func.isRequired,
};

export default Gif;

