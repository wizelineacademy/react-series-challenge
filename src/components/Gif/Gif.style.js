import styled from 'styled-components';

const GifContainer = styled.div`
  float: left;
  height: 300px;
  width: 200px;
  margin: 10px;
  padding: 10px
`;

const GifTitle = styled.p`
  text-align: center;
`;

const GifImage = styled.iframe`
  width: 100%;
  height: 200px;
`;

const GifLikeContainer = styled.img`
  cursor: pointer;
`;

export {
  GifContainer,
  GifTitle,
  GifImage,
  GifLikeContainer,
};
