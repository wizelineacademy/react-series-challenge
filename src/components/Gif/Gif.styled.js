import styled from 'styled-components';
import { Link } from 'react-router-dom';

export const MasonryItem = styled.div`
  display: flex;
  align-items: center;
  flex: auto;
  min-width: 150px;
  margin: 0 8px 8px 0;
  width: ${(props) => props.fullWidth ? '100% !important' : 'initial'};
  justify-content: ${(props) => props.fullWidth ? 'center' : 'initial'};
  &:nth-child(4n+1){
     width: 250px;
  }
  &:nth-child(4n+2){
     width: 325px;
  }
  &:nth-child(4n+3){
     width: 180px;
  }
  &:nth-child(4n+4){
     width: 380px;
  }
  img {
    width: 100%;
  }
`;

export const Overlay = styled.div`
  display: flex;
  justify-content: space-around;
  align-items: center;
  flex-direction: column;
  position:absolute;
  width: 100%;
  height: 100%;
  top: 0;
  right: 0;
  background-color: rgba(0,0,0,0.7);
  opacity: 0;
  transition: 0.5s all;
  &:hover {
    opacity: 1;
  }
`;

export const ImageWrapper = styled.div`
  position:relative;
`;

export const LinkStyled = styled(Link)`
  color: #ffc800;
`;

export const GifImg = styled.img`
  border: ${props => props.isFavorited ? '1px solid #ffc800': 'initial'};
`
export const GifTitle = styled.span`
  color: #ffc800;
  text-align:center;
  visibility: ${props => props.isInDetail ? 'hidden': 'visible'};
`;