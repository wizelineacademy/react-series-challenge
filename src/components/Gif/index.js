import { bindActionCreators } from 'redux';
import { connect } from 'react-redux'
import gif from '../../actions/gif'
import React from 'react'

export const Gif = ({ src, alt, onClick, className, gif, favorite }) => {
    return <img alt={alt} src={src} onClick={onClick.bind(null, gif)} className={ favorite ? `${className} favorite` : className } />
}


const mapDispatchToProps = dispatch => {
    /**
     *
     */
    const { creators: { clickGif: onClick } } = gif
    /**
     *
     */
    return bindActionCreators({ onClick }, dispatch)
}

export default connect(null, mapDispatchToProps)(Gif)