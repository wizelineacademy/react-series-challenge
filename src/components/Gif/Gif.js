import React from 'react';
import {MasonryItem, Overlay, ImageWrapper, LinkStyled, GifImg, GifTitle} from "./Gif.styled";

const getPreviewGifUrl = (gif) => {
  if (gif.images && gif.images.original) {
    return gif.images.original.url;
  } else {
    return '';
  }
};

const handleFavButton = (gif, addGif, removeGif, isFavorited) => {
  if (!isFavorited) {
    addGif(gif);
  } else {
    removeGif(gif.id);
  }
};

export const Gif = ({ gif, fullWidth, addGif, removeGif, isFavorited, isInDetail }) => (
  <MasonryItem fullWidth={fullWidth}>
      <ImageWrapper>
        <GifImg isFavorited={isFavorited} src={getPreviewGifUrl(gif)} alt={gif.title} />
        <Overlay>
          <GifTitle>Title: {gif.title}</GifTitle>
          <button id="fav-btn" onClick={() => handleFavButton(gif, addGif, removeGif, isFavorited)}>
            {!isFavorited ? 'Add to Favorites': 'Remove from favorites'}
          </button>
          { !isInDetail ? (
            <LinkStyled to={{pathname: `/detail/${gif.id}`}}>
              View detail
            </LinkStyled>) : ''
          }
        </Overlay>
      </ImageWrapper>
  </MasonryItem>
)