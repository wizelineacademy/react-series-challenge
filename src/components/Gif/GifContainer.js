import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import favoriteGifsActions from '../../actions/favoriteGifs';
import Gif from './Gif';

const mapDispatchToProps = (dispatch) => {
  const { addFavorite, removeFavorite } = favoriteGifsActions.creators;
  return bindActionCreators({
    addFavorite,
    removeFavorite,
  }, dispatch);
};

export default connect(null, mapDispatchToProps)(Gif);
