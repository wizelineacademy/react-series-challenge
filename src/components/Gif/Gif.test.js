import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Gif } from "./Gif";

describe('Gif component', () => {
  it('should render correctly', () => {
    const output = shallow(
      <Gif
        gif={{ id: '123', title: 'gif' }}
        addGif={jest.fn}
        removeGif={jest.fn}
        fullWidth={false}
        isFavorited={false}
        isInDetail={false}
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });
  it('renders with preview image correctly', () => {
    const output = shallow(
      <Gif
        gif={{ id: '123', title: 'gif', images:{ original: '' } }}
        addGif={jest.fn}
        removeGif={jest.fn}
        fullWidth={true}
        isFavorited={true}
        isInDetail={true}
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });
  it('calls to addGif function if is not favorite', () => {
    const addGif = jest.fn();
    const output = mount(
      <Gif
        gif={{ id: '123', title: 'gif', images:{ original: '' } }}
        addGif={addGif}
        removeGif={jest.fn}
        fullWidth={true}
        isFavorited={false}
        isInDetail={true}
      />
    );
    output.find('button').simulate('click');
    expect(addGif.mock.calls.length).toBe(1);
  });
  it('calls to removeGif function if is favorite', () => {
    const removeGif = jest.fn();
    const output = mount(
      <Gif
        gif={{ id: '123', title: 'gif', images:{ original: '' } }}
        addGif={jest.fn}
        removeGif={removeGif}
        fullWidth={true}
        isFavorited={true}
        isInDetail={true}
      />
    );
    output.find('button').simulate('click');
    expect(removeGif.mock.calls.length).toBe(1);
  });
});