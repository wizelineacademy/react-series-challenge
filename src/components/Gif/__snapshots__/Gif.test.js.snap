// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gif component renders with preview image correctly 1`] = `
<styled.div
  fullWidth={true}
>
  <styled.div>
    <styled.img
      alt="gif"
      isFavorited={true}
      src=""
    />
    <styled.div>
      <styled.span>
        Title: 
        gif
      </styled.span>
      <button
        id="fav-btn"
        onClick={[Function]}
      >
        Remove from favorites
      </button>
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`Gif component should render correctly 1`] = `
<styled.div
  fullWidth={false}
>
  <styled.div>
    <styled.img
      alt="gif"
      isFavorited={false}
      src=""
    />
    <styled.div>
      <styled.span>
        Title: 
        gif
      </styled.span>
      <button
        id="fav-btn"
        onClick={[Function]}
      >
        Add to Favorites
      </button>
      <Styled(Link)
        to={
          Object {
            "pathname": "/detail/123",
          }
        }
      >
        View detail
      </Styled(Link)>
    </styled.div>
  </styled.div>
</styled.div>
`;
